integration-tests/tests/pg.test.ts

```diff
@@ -2384,6 +2387,259 @@ test.serial('select from enum', async (t) => {
+		datetimeString: '2022-01-01T00:00:00.123Z',
+		datetimeFullPrecision: fullPrecision,
+		datetimeWTZString: '2022-01-01T00:00:00.123Z',
+		interval: '1 day',
+	});
+
+	const result = await db.select().from(table);
+
+	Expect<
+		Equal<{
+			id: number;
+			dateString: string;
+			time: string;
+			datetime: Date;
+			datetimeWTZ: Date;
+			datetimeString: string;
+			datetimeFullPrecision: string;
+			datetimeWTZString: string;
+			interval: string;
+		}[], typeof result>
+	>;
+
+	Expect<
+		Equal<{
+			dateString: string;
+			time: string;
+			datetime: Date;
+			datetimeWTZ: Date;
+			datetimeString: string;
+			datetimeFullPrecision: string;
+			datetimeWTZString: string;
+			interval: string;
+			id?: number | undefined;
+		}, typeof table.$inferInsert>
+	>;
+
+	t.deepEqual(result, [
+		{
+			id: 1,
+			dateString: '2022-01-01',
+			time: someTime,
+			datetime: someDatetime,
+			datetimeWTZ: someDatetime,
+			datetimeString: '2022-01-01 00:00:00.123',
+			datetimeFullPrecision: fullPrecision.replace('T', ' ').replace('Z', ''),
+			datetimeWTZString: '2022-01-01 00:00:00.123+00',
+			interval: '1 day',
+		},
+	]);
+
+	await db.execute(sql`drop table if exists ${table}`);
+});
+
+test.serial('all date and time columns with timezone', async (t) => {
+	const { db } = t.context;
+
+	const table = pgTable('all_columns', {
+		id: serial('id').primaryKey(),
+		timestamp: timestamp('timestamp_string', { mode: 'string', withTimezone: true, precision: 6 }).notNull(),
+		timestampAsDate: timestamp('timestamp_date', { withTimezone: true, precision: 3 }).notNull(),
+		timestampTimeZones: timestamp('timestamp_date_2', { withTimezone: true, precision: 3 }).notNull(),
+	});
+
+	await db.execute(sql`drop table if exists ${table}`);
+
+	await db.execute(sql`
+		create table ${table} (
+					id serial primary key,
+					timestamp_string timestamp(6) with time zone not null,
+					timestamp_date timestamp(3) with time zone not null,
+					timestamp_date_2 timestamp(3) with time zone not null
+			)
+	`);
+
+	const timestampString = '2022-01-01 00:00:00.123456-0200';
+	const timestampDate = new Date();
+	const timestampDateWTZ = new Date('2022-01-01 00:00:00.123 +0500');
+
+	const timestampString2 = '2022-01-01 00:00:00.123456-0400';
+	const timestampDate2 = new Date();
+	const timestampDateWTZ2 = new Date('2022-01-01 00:00:00.123 +0200');
+
+	await db.insert(table).values([
+		{ timestamp: timestampString, timestampAsDate: timestampDate, timestampTimeZones: timestampDateWTZ },
+		{ timestamp: timestampString2, timestampAsDate: timestampDate2, timestampTimeZones: timestampDateWTZ2 },
+	]);
+
+	const result = await db.select().from(table);
+	const result2 = await db.execute<{
+		id: number;
+		timestamp_string: string;
+		timestamp_date: string;
+		timestamp_date_2: string;
+	}>(sql`select * from ${table}`);
+
+	// Whatever you put in, you get back when you're using the date mode
+	// But when using the string mode, postgres returns a string transformed into UTC
```

Comment:
We have a few requests to make changes in the current PR.

Could you please create separate test cases for the different blocks I've sent you? You should have this set of tests:
```ts
test.serial('Test mode string for timestamp with timezone', async (t) => {
    // 1. Insert date in string format with timezone in it
    // 2. Select in string format and check that values are the same
    // 3. Select as raw query and check that values are the same
    ...
})

test.serial('Test mode date for timestamp with timezone', async (t) => {
...
})

// and so on
```

It would be beneficial to separate tests into logical blocks. This practice aids in reviewing and identifying broken tests more efficiently. It shouldn't consume much of your time. Before approval, you can focus on `pg.test.ts` only, eliminating the need for multiple copy-pastes

---

Additionally, add one more test case in string mode. The reason you are receiving timestamps converted to UTC from Postgres is that it appears your PostgreSQL timezone is set to 'UTC,' causing Postgres to return timestamps converted to the database's own timezone

Here is a quote from Postgres docs:
> In a literal that has been determined to be timestamp without time zone, PostgreSQL will silently ignore any time zone indication. That is, the resulting value is derived from the date/time fields in the input value, and is not adjusted for time zone.

> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's [TimeZone](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-TIMEZONE) parameter, and is converted to UTC using the offset for the timezone zone.

So, to structure this test properly, you need to add a few actions and create another test with the same actions:

Test 1 for mode string

1. Run SET TIME ZONE '<timezone you want to test with>'. Let's set this test to UTC, so it will be `SET TIME ZONE 'UTC'`
2. Run all tests as usual.
3. Return the database to the default timezone.

Test 2 for mode string
1. Run SET TIME ZONE '<timezone you want to test with>'. Let's set this test to the timezone you used when inserting the date.
2. Run all tests as usual.
3. Return the database to the default timezone.
> You can use SELECT * FROM pg_timezone_names; to get offsets for each timezone name and rely on it."

Reply:
Please let me know upon reviewing if you'd like me to add the tests to the rest of the pg drivers.

Reply:
@Angelelz 1 thing left is to separate test cases for the different blocks, same as they are separated [here](https://github.com/drizzle-team/drizzle-orm/pull/1659#issuecomment-1871033758)

So `all date and time columns with timezone` should be separated into 3 tests: 

1. **First case(mode: string)**
Insert date in string format with timezone in it
Select in string format and check that values are the same
Select as raw query and check that values are the same

2. **Second case(mode: date)**
Insert date as new Date()
Select as date and check that timezones are the same
Compare both dates(maybe by using .getTime())

3. **Third case(mode: date) with timestamps inserted from different timezones:**
Insert date as new Date() with one timezone
Insert date as new Date() with another timezone
Select both dates and compare both dates(maybe by using .getTime())

And `all date and time columns without timezone` should be separated into 3 tests:

1. **First case(mode: string)**
Insert date in string format without timezone
Select in string format and check that values are the same
Select as raw query and check that values are the same

2. **Second case(mode: string)**
Insert date in string format with timezone
Select as raw query and check that values are the same
Example: You have inserted 2023-12-28T09:29:43.824+03:00, You selected 2023-12-28 09:29:43.824. You need to trim T and timezone from the first date and check that values are the same

3. **Third case(mode: date)**
Insert date as new Date()
Select as raw query as string, make this as Date in UTC and compare both dates(maybe by using .getTime())
Compare both dates using orm mapping(maybe by using .getTime())

Reply:
Sorry, I didn't read this properly initially, I thought you only wanted those additional tests. No problem, this should be an easy one.